lib/ReactViews/Feedback/FeedbackForm.tsx

```diff
@@ -0,0 +1,454 @@
+import Text from "../../Styled/Text";
+import parseCustomMarkdownToReact, {
+  parseCustomMarkdownToReactWithOptions
+} from "../Custom/parseCustomMarkdownToReact";
+import { useTranslationIfExists } from "./../../Language/languageHelpers";
+
+interface IProps extends WithTranslation {
+  theme: DefaultTheme;
+  viewState: ViewState;
+  t: TFunction;
+}
+
+interface IState {
+  name: string;
+  email: string;
+  comment: string;
+  isSending: boolean;
+  sendShareURL: boolean;
+  commentIsValid: boolean;
+}
+
+@observer
+class FeedbackForm extends React.Component<IProps, IState> {
+  static displayName = "FeedbackForm";
+
+  state: IState = {
+    isSending: false,
+    sendShareURL: true,
+    name: "",
+    email: "",
+    comment: "",
+    commentIsValid: false
+  };
+  escKeyListener: (e: any) => void;
+
+  constructor(props: IProps) {
+    super(props);
+    this.escKeyListener = e => {
+      if (e.keyCode === 27) {
+        this.onDismiss();
+      }
+    };
+    this.onDismiss = this.onDismiss.bind(this);
+    this.updateName = this.updateName.bind(this);
+    this.updateEmail = this.updateEmail.bind(this);
+    this.updateComment = this.updateComment.bind(this);
+    this.onSubmit = this.onSubmit.bind(this);
+    this.changeSendShareUrl = this.changeSendShareUrl.bind(this);
+  }
+
+  getInitialState() {
+    return {
+      isSending: false,
+      sendShareURL: true,
+      name: "",
+      email: "",
+      comment: ""
+    };
+  }
+
+  componentDidMount() {
+    window.addEventListener("keydown", this.escKeyListener, true);
+    this.setState({
+      commentIsValid:
+        this.props.viewState.terria.configParameters.feedbackMinLength === 0
+    });
+  }
+
+  componentWillUnmount() {
+    window.removeEventListener("keydown", this.escKeyListener, true);
+  }
+
+  resetState() {
+    this.setState(this.getInitialState());
+  }
+
+  onDismiss() {
+    runInAction(() => {
+      this.props.viewState.feedbackFormIsVisible = false;
+    });
+    this.resetState();
+  }
+
+  updateName(e: React.ChangeEvent<HTMLInputElement>) {
+    this.setState({
+      name: e.target.value
+    });
+  }
+
+  updateEmail(e: React.ChangeEvent<HTMLInputElement>) {
+    this.setState({
+      email: e.target.value
+    });
+  }
+
+  updateComment(e: React.ChangeEvent<HTMLTextAreaElement>) {
+    this.setState({
+      comment: e.target.value
+    });
+    if (
+      this.state.comment.replace(/\s+/g, " ").length >=
+      this.props.viewState.terria.configParameters.feedbackMinLength!
+    ) {
+      this.setState({
+        commentIsValid: true
+      });
+    } else {
+      this.setState({
+        commentIsValid: false
+      });
+    }
+  }
+
+  changeSendShareUrl(e: React.ChangeEvent<HTMLInputElement>) {
+    this.setState((prevState: IState) => ({
+      sendShareURL: !prevState.sendShareURL
+    }));
+  }
+
+  onSubmit(e: React.FormEvent<HTMLFormElement | HTMLDivElement>) {
+    e.preventDefault();
+
+    if (
```

Comment:
instead of detecting this onSubmit, we could instead set the minLength attribute on the textarea. Since the textarea has a name attribute, this will trigger the browser's form validation and we also get a nice warning message on why the submit failed.